Runtime call is added to pennylane (#9970)
PennyLaneAI/pennylane: doc/releases/changelog-dev.md

  decomposed recursively into :class:`~.FermionicSWAP` and :class:`~.TwoWireFFT` operations
  (two-site Fermionic Fourier transforms).

> * A new `pennylane.backline.runtime` module is added, it lets a compiled program call a runtime entry point directly, by its C symbol name. A symbol's signature is declared once with `qp.runtime_declare` and called with `qp.runtime_call` from inside a `qjit` program.

Reviewer: ```suggestion
* A new `pennylane.backline.runtime` module is added, adding the ability for a compiled program to call a runtime entry point directly via its C symbol name. A symbol's signature is declared once with `qp.runtime_declare` and called with `qp.runtime_call` from inside a `qjit` program.
```

@rniczh worth also mentioning that this is only supported for programs with `qnode(backline)`? (I assume it won't work for a general qjit --- let me know if this is not the case!)

Author: It doesn't require the qnode and even backline actually, for example: 

The remote runtime call can be executed like this:
```python
def call(op, *args, at, out_bytes=0):
    """One dispatched transport call, on the executor at ``at``."""
    return qp.runtime_call(f"{PREFIX}{op}__wrapper", *args, address=at, out_bytes=out_bytes)
     
...
    
@qjit
def remote_benchmark():
    co = call("create", COPROC_LIB, COPROC_CFG, ROLE_COPROCESSOR, KEY, at=coproc)
    ct = call("create", CTRL_LIB, CTRL_CFG, ROLE_CONTROLLER, KEY, at=ctrl)
    ...
```

So we just need qjit for compiling. And for the local runtime_call, it could be just a pure CAPI function call to call any function defined in libraries that can be searched from compiled executable. So both qnode and backline are not necessary.

Reviewer: oh awesome! In that case, sounds good

  [(#9970)](https://github.com/PennyLaneAI/pennylane/pull/9970)

  ```python
  import pennylane as qp

  qp.runtime_declare("example_run_rounds", "(ptr, u32) -> u64", library="/path/librounds.so")

  def program(session):
      return qp.runtime_call("example_run_rounds", session, 100000)
  ```

  Passing `address="host:port"` dispatches the call to executor on remote side, which invokes the

>   symbol on the machine the runtime lives on; without it the call is local. And `qp.backline.runtime.CType` lists what can cross the boundary

Reviewer: ```suggestion
  symbol on the machine the runtime lives on; without it the call is local. Meanwhile, `qp.backline.runtime.CType` lists what can cross the boundary.
```

  The nodes are taken from the placement of the device being traced, so a round on a built backline is just `correction = qp.backline.decode(syndrome)`:

  ```python
  import pennylane as qp

  con = qp.Controller(
      device=qp.device("lightning.qubit", wires=2),
      remote=True,
      executor_options={"host": address_controller},
      init_args={"out_bytes": 8},
  )
  coproc = qp.Coprocessor(
      coprocessor_fn="decoder",
      label="decoder-0",
      backend="gpu_verbs",
      comm_host=address_coproc,
      oob_port=18590,
  )
  dev = qp.backline(controller=con, coprocessors=[coproc], transport="rdma")

  @qp.qjit(capture=True)
  @qp.qnode(dev)
  def circuit(syndrome):
      correction = qp.backline.decode(syndrome)
      qp.cond(correction[0] == 1, qp.X)(0)
      return qp.expval(qp.Z(0))
  ```

  The round is resolved from the device being traced, so the program has to be captured (`qp.qjit(capture=True)`). The correction comes back as a `uint8` buffer of the controller's committed `out_bytes`. Pass `controller=` / `coprocessor=` to choose the nodes explicitly, and `decoder_id=` to select which coprocessor-side decoder handles the round.

> 

Reviewer: nice entry!

  decomposed recursively into :class:`~.FermionicSWAP` and :class:`~.TwoWireFFT` operations
  (two-site Fermionic Fourier transforms).

> * A new `pennylane.backline.runtime` module is added, adding the ability for a compiled program to call a runtime entry point directly via its C symbol name. A symbol's signature is declared once with `qp.runtime_declare` and called with `qp.runtime_call` from inside a `qjit` program.

Reviewer: ```suggestion
* The ability for a compiled program to call a runtime entry point directly via its C symbol name has been added. A symbol's signature is declared once with `qp.runtime_declare` and called with `qp.runtime_call` from inside a `qjit` program.
```

Since the module is just an implementation detail, can leave it out of the release notes

      return reply
  ```

* A new `qp.backline.decode` function is added, which offloads one syndrome to a coprocessor from inside a captured QNode and returns its correction, driving a single transport round: stage the syndrome, post it, collect the reply.
  [(#9970)](https://github.com/PennyLaneAI/pennylane/pull/9970)

  The nodes are taken from the placement of the device being traced, so a round on a built backline is just `correction = qp.backline.decode(syndrome)`:

  ```python
  import pennylane as qp

  address_controller = "192.168.3.15"  # the host the controller runs on
  address_coproc = "192.168.1.3"  # the host the coprocessor listens on

  con = qp.Controller(
      device=qp.device("null.qubit", wires=2),
      executor_options={"host": address_controller},
      init_args={"out_bytes": 8},
  )
  coproc = qp.Coprocessor(
      coprocessor_fn="decoder",
      label="decoder-0",
      backend="gpu_verbs",
      comm_host=address_coproc,
      oob_port=18590,
  )
  dev = qp.backline(controller=con, coprocessors=[coproc], transport="rdma")

  @qp.qjit(capture=True)
  @qp.qnode(dev)
  def circuit(syndrome):
      correction = qp.backline.decode(syndrome)

>       qp.cond(correction[0] == 1, qp.X)(0)

Reviewer: @rniczh just checking, does this also work with autograph?

```python
  @qp.qjit(capture=True, autograph=True)
  @qp.qnode(dev)
  def circuit(syndrome):
      correction = qp.backline.decode(syndrome)
      if correction[0] == 1:
          qp.X(0)
      return qp.expval(qp.Z(0))
```

Author: Yes, just tested it, it did generate all the instructions we need: (removed other instructions in between)

```mlir
%2 = catalyst.custom_call fn("__catalyst__transport__get_session__call") (%1, %arg0) {number_original_arg = 2 : i32} : (tensor<1xi32>, tensor<256xui8>) -> tensor<1xui64>
%8 = catalyst.custom_call fn("__catalyst__transport__stage_payload__call") (%5, %arg1, %6, %7) {number_original_arg = 4 : i32} : (tensor<1xui64>, tensor<4xui8>, tensor<1xui64>, tensor<1xui32>) -> tensor<1xi32>
%11 = catalyst.custom_call fn("__catalyst__transport__post__call") (%9, %10) {number_original_arg = 2 : i32} : (tensor<1xui64>, tensor<1xui32>) -> tensor<1xi32>
%14:2 = catalyst.custom_call fn("__catalyst__transport__collect__call") (%12, %13) {number_original_arg = 2 : i32} : (tensor<1xui64>, tensor<1xui64>) -> (tensor<1xi32>, tensor<8xui8>)  
%extracted = tensor.extract %17[] : tensor<i1>
scf.if %extracted {
  %21 = qref.get %0[ 0] : !qref.reg<2> -> !qref.bit
  qref.custom "PauliX"() %21 : !qref.bit
}
```

  decomposed recursively into :class:`~.FermionicSWAP` and :class:`~.TwoWireFFT` operations
  (two-site Fermionic Fourier transforms).

* A new `pennylane.backline.runtime` module is added, adding the ability for a compiled program to call a runtime entry point directly via its C symbol name. A symbol's signature is declared once with `qp.runtime_declare` and called with `qp.runtime_call` from inside a `qjit` program.
  [(#9970)](https://github.com/PennyLaneAI/pennylane/pull/9970)

  ```python
  import pennylane as qp

  qp.runtime_declare("example_local_rounds", "(ptr, u32) -> u64", library="/path/librounds.so")

  def program(session):
      return qp.runtime_call("example_local_rounds", session, 100000)
  ```

>   Passing `address="host:port"` dispatches the call to executor on remote side, which invokes the

Reviewer 2: ```suggestion
  Passing `address="host:port"` dispatches the call to the executor on the remote side, which invokes the
```
